rescript/ncbi_datasets.py

```diff
@@ -0,0 +1,159 @@
+# ----------------------------------------------------------------------------
+
+import glob
+import os
+import shutil
+import tempfile
+import zipfile
+from typing import List
+
+import ncbi.datasets as nd
+from ncbi.datasets.openapi import ApiClient as DatasetsApiClient
+import pandas as pd
+import skbio
+from multiprocessing import Manager
+from ncbi.datasets import ApiException
+from q2_types.feature_data import DNAFASTAFormat
+from q2_types_genomics.genome_data import (LociDirectoryFormat,
+                                           ProteinsDirectoryFormat)
+
+from rescript.ncbi import get_taxonomies, _default_ranks
+
+
+def _get_assembly_descriptors(
+        api_instance, assembly_levels, assembly_source, only_reference,
+        page_size, taxon, tax_exact_match
+):
+    all_acc_ids = []
+    all_tax_ids = []
+    next_page_token = ''
+    while True:
+        try:
+            genome_summary = api_instance.assembly_descriptors_by_taxon(
+                taxon=taxon,
+                page_size=page_size,
+                filters_assembly_source=assembly_source,
+                filters_assembly_level=assembly_levels,
+                filters_reference_only=only_reference,
+                # TODO: possibly, make this configurable - probably will
+                #   require handling some genomes without annotation
+                #   (maybe an issue for loci/proteins types?)
+                filters_has_annotation=True,
+                tax_exact_match=tax_exact_match,
+                page_token=next_page_token
+            )
+        except ApiException as e:
+            raise Exception('Unexpected error while calling NCBI Datasets '
+                            f'API: {e}.')
+
+        if not genome_summary.assemblies:
+            raise Exception('The query to the NCBI Dataset API did not return '
+                            'any result but contained an error message (see '
+                            'below). Please check the message, update your '
+                            'query and try again.\n\nThe error message was:\n'
+                            f'"{genome_summary.messages[0].error.message}"')
+
+        next_page_token = genome_summary.next_page_token
+
+        genome_assembly = [x.assembly for x in genome_summary.assemblies]
+
+        all_acc_ids.extend([x.assembly_accession for x in genome_assembly])
+        all_tax_ids.extend([x.org.tax_id for x in genome_assembly])
+
+        if not next_page_token:
+            break
+    return all_acc_ids, all_tax_ids
+
+
+def _fetch_and_extract_dataset(api_response):
+    with tempfile.TemporaryDirectory() as tmp:
+        result_path = os.path.join(tmp, 'datasets.zip')
+        with open(result_path, 'wb') as f:
+            f.write(api_response.data)
+
+        with zipfile.ZipFile(result_path, 'r') as zipped:
+            zipped.extractall(tmp)
+
+        # find and move all the genome sequences
+        genomes = DNAFASTAFormat()
```

Comment:
Should be able to use `MixedCaseDNAFASTAFormat` as outlined [here](https://forum.qiime2.org/t/importing-sequence-data-with-lower-case-nucleotide-characters/25158). This will convert the sequences to upper case for you.

Reply:
Unfortunately, this will not help, as sequences need to be uppercase already when reading by `skbio.read` here 😢: https://github.com/misialq/RESCRIPt/blob/eab69f0b7fc903212aca433859a03db7d9944d6e/rescript/ncbi_datasets.py#L91-L93